src/main/java/org/jabref/logic/importer/fetcher/NewAstrophysicsDataSystem.java

```diff
@@ -0,0 +1,266 @@
+    public Parser getParser() {
+        return new BibtexParser(preferences, new DummyFileUpdateMonitor());
+    }
+
+    @Override
+    public void doPostCleanup(BibEntry entry) {
+        new FieldFormatterCleanup(StandardField.ABSTRACT, new RemoveBracesFormatter()).cleanup(entry);
+        new FieldFormatterCleanup(StandardField.ABSTRACT, new RemoveNewlinesFormatter()).cleanup(entry);
+        new FieldFormatterCleanup(StandardField.TITLE, new RemoveBracesFormatter()).cleanup(entry);
+        new FieldFormatterCleanup(StandardField.AUTHOR, new NormalizeNamesFormatter()).cleanup(entry);
+
+        // Remove ADS note
+        new FieldFormatterCleanup(new UnknownField("adsnote"), new ClearFormatter()).cleanup(entry);
+        // Move adsurl to url field
+        new MoveFieldCleanup(new UnknownField("adsurl"), StandardField.URL).cleanup(entry);
+        // The fetcher adds some garbage (number of found entries etc before)
+        entry.setCommentsBeforeEntry("");
+    }
+
+    @Override
+    public List<BibEntry> performSearch(BibEntry entry) throws FetcherException {
+
+        if (entry.getFieldOrAlias(StandardField.TITLE).isEmpty() && entry.getFieldOrAlias(StandardField.AUTHOR).isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        try {
+            List<String> bibcodes = fetchBibcodes(getURLForEntry(entry));
+            String[] bibcodeArray = new String[bibcodes.size()];
+            bibcodes.toArray(bibcodeArray);
+            return performSearchByIds(bibcodeArray);
+        } catch (URISyntaxException e) {
+            throw new FetcherException("Search URI is malformed", e);
+        } catch (IOException e) {
+            throw new FetcherException("A network error occurred", e);
+        }
+    }
+
+    @Override
+    public List<BibEntry> performSearch(String query) throws FetcherException {
+
+        if (StringUtil.isBlank(query)) {
+            return Collections.emptyList();
+        }
+
+        try {
+            List<String> bibcodes = fetchBibcodes(getURLForQuery(query));
+            String[] bibcodeArray = new String[bibcodes.size()];
+            return performSearchByIds(bibcodes.toArray(bibcodeArray));
+        } catch (URISyntaxException e) {
+            throw new FetcherException("Search URI is malformed", e);
+        } catch (IOException e) {
+            throw new FetcherException("A network error occurred", e);
+        }
+    }
+
+    private List<String> fetchBibcodes(URL url) throws FetcherException {
+
+        try {
+            URLDownload download = new URLDownload(url);
+            download.addHeader("Authorization", "Bearer " + API_KEY);
+            String content = download.asString();
+            JSONObject obj = new JSONObject(content);
+
+            try {
+                JSONArray codes = obj.getJSONObject("response").getJSONArray("docs");
+                List<String> bibcodes = new ArrayList<>();
+                for (int i = 0; i < codes.length(); i++) {
```

Comment:
You really have to be careful to check if all properties of the json response are mandatory or if they can be optional. In the latter case you will run into exceptions if you don't check if it has the attribute

Reply:
I checked the API. I found no indication on their resource pages that the field I used was optional. I tested several edge cases on the endpoint and no result was missing these fields or if they were missing, they returned 404 anyway.
And I catch JSONException and return an empty list.